server/apiserver/argoserver.go

```diff
@@ -325,16 +326,21 @@ func (as *argoServer) newGRPCServer(instanceIDService instanceid.Service, offloa
 // using grpc-gateway as a proxy to the gRPC server.
 func (as *argoServer) newHTTPServer(ctx context.Context, port int, artifactServer *artifacts.ArtifactServer) *http.Server {
 	endpoint := fmt.Sprintf("localhost:%d", port)
+	ipKeyFunc := httplimit.IPKeyFunc()
+	if ipKeyFuncHeadersStr := env.GetString("IP_KEY_FUNC_HEADERS", ""); ipKeyFuncHeadersStr != "" {
+		ipKeyFuncHeaders := strings.Split(ipKeyFuncHeadersStr, ",")
+		ipKeyFunc = httplimit.IPKeyFunc(ipKeyFuncHeaders...)
```

Comment:
This IPKeyFunc(https://github.com/sethvargo/go-limiter/blob/v0.7.2/httplimit/middleware.go#L45) will return errors if the ip is not valid. 

I am okay with this behaviour because it's the users fault for supplying bad IPs. 

Just wanted to point this out if you think this is acceptable or not.

Reply:
the `IPKeyFunc` was already used before as well, just with no arguments passed in, so we've already been returning errors in that case. haven't really seen problems with it, so I'm inclined to not change that

> I am okay with this behaviour because it's the users fault for supplying bad IPs.

so that depends on who adds the header -- usually it's a proxy, client, etc. It's not necessarily the "user" in that sense.

but that is generally an input problem / bug in the user's software and I do think it's ok to continue erroring in that case.
our other options are to return a more specific error or to bucket all IP address parse failures together into one bucket for the purpose of rate limitng